### PR TITLE
add activation codgen with torch-mlir converter

### PR DESCRIPTION
**Pull request by Yancey1989**

### CODE REVIEW

File: pytorch_blade/tests/disc/ops/test_unary_ops.py [hunk 1]
```diff
@@ -71,23 +71,23 @@ def tanh_func(x):
 
         self._test_unary_ops(tanh_func)
 
-    @skipIfNoTorchMlir()
+    @skipIfEnableTorchMlir()
```
Comment (fortianyou) [inline pytorch_blade/tests/disc/ops/test_unary_ops.py:74]: Nit: need rebase

File: pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/TorchToMhlo.cpp [hunk 2]
```diff
@@ -525,25 +525,87 @@ LogicalResult ConvertAtenOp<AtenTanhOp>::matchAndRewrite(
...
-  } else {
-    // Sigmoid legalization in MHLO for quantized element-type uses
-    // specialized mhlo.table construct.
-    return op.emitError(
-        "Only floating-point datatype legalization currently supported");
-  }
+  Value one = chlo::getConstantLike(rewriter, loc, 1.0, self);
+  Value negVal = rewriter.create<mhlo::NegOp>(loc, self);
+  Value expVal = rewriter.create<mhlo::ExpOp>(loc, negVal);
+  Value addVal = rewriter.create<mhlo::AddOp>(loc, expVal, one);
```
Comment (fortianyou) [inline pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/TorchToMhlo.cpp:591]: Nit: is the broadcast needed to be considered?

File: pytorch_blade/tests/disc/ops/test_unary_ops.py [hunk 3]
```diff
@@ -71,23 +71,23 @@ def tanh_func(x):
...
     def test_contiguous(self):
         @torch.jit.script
         def contiguous_func(x):
             return x.contiguous()
 
         self._test_unary_ops(contiguous_func)
 
-    @skipIfNoTorchMlir()
+    @skipIfEnableTorchMlir()
```
Comment (fortianyou) [inline pytorch_blade/tests/disc/ops/test_unary_ops.py:82]: export TORCH_DISC_USE_TORCH_MLIR=true needs to be set to enable the TorchMLIR UTs

File: pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/TorchToMhlo.cpp [hunk 4]
```diff
@@ -525,25 +525,87 @@ LogicalResult ConvertAtenOp<AtenTanhOp>::matchAndRewrite(
...
+// Convert a Aten::Relu to HLO
+// Relu(x) = 0 if x < 0 else x
+template <>
+LogicalResult ConvertAtenOp<AtenReluOp>::matchAndRewrite(
+    AtenReluOp op,
+    OpAdaptor adaptor,
+    ConversionPatternRewriter& rewriter) const {
+  Location loc = op.getLoc();
+  Value input = adaptor.self();
+  auto inputTy = input.getType().cast<TensorType>();
```
Comment (wyzero) [inline pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/TorchToMhlo.cpp:537]: Is AtenReluOp always be ranked? maybe we need to check is the input type is RankedTensorType?

File: pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/TorchToMhlo.cpp [hunk 5]
```diff
@@ -525,25 +525,87 @@ LogicalResult ConvertAtenOp<AtenTanhOp>::matchAndRewrite(
...
+      loc, input, zero, mhlo::ComparisonDirection::GT);
+  rewriter.replaceOpWithNewOp<mhlo::SelectOp>(
+      op, inputTy, compareGtZero, input, zero);
+  return success();
+}
+
+// Convert a Aten::LeakyRelu to HLO
+// LeakyRelu(x) = max(0, x) + negative_slop * min(0, x)
+template <>
+LogicalResult ConvertAtenOp<AtenLeakyReluOp>::matchAndRewrite(
```
Comment (wyzero) [inline pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/TorchToMhlo.cpp:549]: nit: FileCheck based UT or open an issue to track this TODO?.

File: pytorch_blade/tests/disc/ops/test_activation.py [hunk 6]
```diff
@@ -11,65 +11,49 @@
...
-        def gelu_func(x):
-            return torch.nn.functional.gelu(x)
-        self._test_activation(gelu_func)
+        self._test_activation(torch.nn.GELU(), torch.nn.functional.gelu)
+
+    @skipIfEnableTorchMlir()
+    def test_hardtanh(self):
+        self._test_activation(torch.nn.Hardtanh(), torch.nn.functional.hardtanh)
 
+    @skipIfEnableTorchMlir()
```
Comment (wyzero) [inline pytorch_blade/tests/disc/ops/test_activation.py:56]: What's purpose of this decorator?

Comment (wyzero) [inline pytorch_blade/tests/disc/ops/test_activation.py:56]: What's purpose of this decorator?

File: pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/TorchToMhlo.cpp (hunk 5 above)
Comment (Yancey1989) [reply pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/TorchToMhlo.cpp:549]: Thanks, I created an issue https://github.com/alibaba/BladeDISC/issues/385 to trace this task.

File: pytorch_blade/tests/disc/ops/test_activation.py (hunk 6 above)
Comment (Yancey1989) [reply pytorch_blade/tests/disc/ops/test_activation.py:56]: No all ut functions are supported in torch-mlir converter, I just use this decorator to pass these convertor ut that are not implemented.

File: pytorch_blade/tests/disc/ops/test_unary_ops.py (hunk 3 above)
Comment (Yancey1989) [reply pytorch_blade/tests/disc/ops/test_unary_ops.py:81]: The `unittest` would run all functions with the prefix `test`, actually, we should SKIP some of them if not implemented in torch-mlir converter.

**Comment by Yancey1989** [comment]
After offline discussion, we should check the converter with static shape, full dynamic shape, and partial dynamic shape, so I will fix #383 before continuing this PR.

File: pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/TorchToMhlo.cpp (hunk 2 above)
Comment (Yancey1989) [reply pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/TorchToMhlo.cpp:621]: I think `chlo::getConstantLike` works well, part of the IR:
```text
    %1 = "chlo.constant_like"(%0) {value = 1.000000e+00 : f32} : (tensor<?x?x?x?xf32>) -> tensor<?x?x?x?xf32>
```

File: pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/TorchToMhlo.cpp (hunk 4 above)
Comment (Yancey1989) [reply pytorch_blade/src/torch-mlir/lib/Conversion/TorchToMhlo/TorchToMhlo.cpp]: done

File: pytorch_blade/tests/disc/ops/test_unary_ops.py (hunk 1 above)
Comment (Yancey1989) [reply pytorch_blade/tests/disc/ops/test_unary_ops.py]: done

**Review by fortianyou — APPROVED** [review]
LGTM

**Pull request merged**